### PR TITLE
Convert codebase to TypeScript

### PR DESCRIPTION
**Pull request by gyoshev**
## Background

This PR converts the codebase to TypeScript and updates the build process to handle this.
It also generates type declarations that are published along with the NPM package.

Note: Unit tests are still written in JS.

## Implementation

- Replace Babel with webpack ts-loader
- Add types to all public methods
- Output type definitions in the `dist/leanplum.d.ts`

## Testing strategy

`yarn test` and ran e2e test with the generated 

## Backwards compatibility

Compatible, the only changed code is a removed noop in

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -7,42 +7,37 @@
   },
   "license": "Apache-2.0",
   "dependencies": {
-    "babel-eslint": "^7.2.3",
+    "@types/node": "^13.5.0",
```
Comment (iignatov) [inline package.json]: nit: IMO this should be moved to `devDependencies`.

File: dist/leanplum.d.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            * Unregisters the browser form web push.
+            * @return {Promise}            Resolves on success, otherwise rejects.
+            */
+        static unregisterFromWebPush(): Promise<string>;
+        /**
+            * Clears cached values for messages, variables and test assignments.
+            * Use sparingly as if the app is updated, you'll have to deal with potentially
+            * inconsistent state or user experience.
+            */
+        static clearUserContent(): void;
```
Comment (iignatov) [inline dist/leanplum.d.ts]: nit: I'm not sure if this is auto-generated but it seems that there's too much indentation (and misaligned comments).

File: test/e2e/js/index.js [hunk 3]
```diff
@@ -1,54 +1,37 @@
...
-      exports: {}
-    };
-    factory();
-    global.indexEs6 = mod.exports;
-  }
-})(this, function () {
-  'use strict';
+/*
+ *
+ *  Copyright 2017 Leanplum Inc. All rights reserved.
```
Comment (iignatov) [inline test/e2e/js/index.js]: nit:
```suggestion
 *  Copyright 2020 Leanplum Inc. All rights reserved.
```

File: dist/leanplum.d.ts (hunk 2 above)
Comment (gyoshev) [reply dist/leanplum.d.ts]: It's auto-generated from [this code](https://github.com/Leanplum/Leanplum-JavaScript-SDK/blob/8d1c4909aedf35e549f34d293545d248fd773ace/src/Leanplum.ts#L389-L392), and it seems that the generation doesn't align the comments.

**Pull request merged**